src/object/cli_obj.c

```diff
@@ -3930,6 +3930,7 @@ obj_comp_cb(tse_task_t *task, void *data)
 				obj_auxi_free_failed_tgt_list(obj_auxi);
 				obj_reasb_req_fini(&obj_auxi->reasb_req,
 						   obj_auxi->iod_nr);
+				obj_bulk_fini(obj_auxi);
```

Comment:
So this bulk_fini is for releasing the bulk in the original fetch task? then this should be done after obj_ec_recov_cb(), in case recov_cb() failed, recover task never be scheduled()?  @liuxuezhao

Reply:
here the data recovery has been finished and recovered data copied to original user buffer.
This is a hg bulk handle leak, the fix looks good to me. below memset(obj_auxi, 0, sizeof(*obj_auxi)) zeroed everyting.

obj_ec_recov_cb is start data recovery, and obj_ec_recov_data is to finish the data recovery.

Reply:
data recovery fetch is done by the sub task created by obj_ec_recov_cb(), and its bulk will be freed in the following else {}.

So this bulk free is for the original fetch() ? then it should be done after obj_ec_recov_cb(), in case there are any failure there.

Reply:
"So this bulk free is for the original fetch() ?" yes

"then it should be done after obj_ec_recov_cb(), in case there are any failure there."
if some sub-task failed, then the failure should be able to reflect to parent task (original fetch task, which with "ec_wait_recov" flag)'s completion cb - obj_comp_cb().

Reply:
hmm, I checked some code, that looks better to add a few other err handing in obj_ec_recov_cb() or obj_ec_recov_cb()'s caller for the case that it failed without any sub-task created. I'll refine it in one of my other patch. But the bulk handle cleanup here seems fine.

Reply:
because we memset() in this code block, so if we don't release bulk handle, it is definitely a memory leak.

Reply:
>if some sub-task failed, then the failure should be able to reflect to parent task (original fetch task, which with "ec_wait_recov" flag)'s completion cb - obj_comp_cb().


Only if the task is rescheduled successfully? otherwise the bulk will be leaked?

Reply:
>because we memset() in this code block, so if we don't release bulk handle, it is definitely a memory leak.

I mean you should do it after obj_ec_recov_cb() inside else { }

Reply:
I just refreshed the patch, do the bulk_fini at both place.
I'll refine some other details err handling in other following patch.